Dev

uladzislau-zhurauski opened the pull request.

uladzislau-zhurauski committed:
Create models and migrations

uladzislau-zhurauski committed:
Set up admin

uladzislau-zhurauski committed:
Add image check

uladzislau-zhurauski committed:
Refactor foreign keys, add new model ProductMaterial, remove commented code

uladzislau-zhurauski committed:
Add some improvements

Set all max_length options to 255 except for the phone_number field
Add upload_to option to image
Set media_url and media_root in settings
Generate tip field of Image in save method

uladzislau-zhurauski committed:
Edit readme

uladzislau-zhurauski committed:
Add troubleshooting section to readme

uladzislau-zhurauski committed:
Merge branch 'OS-2__Create_README_and_gitignore' into dev

uladzislau-zhurauski committed:
Create basic product retrieving functionality

uladzislau-zhurauski committed:
Refactor product functionality according to mvc

uladzislau-zhurauski committed:
Refactor controller code

uladzislau-zhurauski committed:
Some refactoring related to mvc pattern

uladzislau-zhurauski committed:
Restructure image model with generic relations

uladzislau-zhurauski committed:
Fix a small typo

uladzislau-zhurauski committed:
Improve the message in the exception

uladzislau-zhurauski committed:
Add feedback backend functionality

Add permissions
Add custom manager

uladzislau-zhurauski committed:
Add todo and remove unnecessary blank lines

uladzislau-zhurauski committed:
Fix feedback_detail put and delete permissions

uladzislau-zhurauski committed:
Create test environment, views tests, etc.

Update .gitignore (duplicate pattern), settings.py (no template dir needed)
Modify product and feedback serializers and their views to run tests
Fix response in successful feedback put
Delete shop/views.py

uladzislau-zhurauski committed:
Add instruction for running tests to readme

uladzislau-zhurauski committed:
Add filling the test database at the beginning of the test session

uladzislau-zhurauski committed:
Add models tests, factories, etc

Add additional factories and move all factories into one file
Add parent_category field to category factory
Change content_type_choices from generator to list comprehension

uladzislau-zhurauski committed:
Add generic non-existent pk to conftest

uladzislau-zhurauski committed:
Refactor feedback tests so that they use a new multi_client fixture

Combine authenticated_api_client and authenticated_api_staff_client into one authenticated_api_client fixture
Rename some tests

uladzislau-zhurauski committed:
Add image backend

uladzislau-zhurauski committed:
Add ImageInline to product and feedback admin

uladzislau-zhurauski committed:
Move check_permissions decorator

uladzislau-zhurauski committed:
Add image save tests

uladzislau-zhurauski committed:
Add generic existent pk to conftest

uladzislau-zhurauski committed:
Add image views tests

Add file fixtures in conftest
Modify ImageFactory image field

uladzislau-zhurauski committed:
Refactor models str tests and image save test

Modify image factories
Add creation of different images to test data

uladzislau-zhurauski committed:
Add image uploading to feedback backend

Add image and product serializers to feedback serializer

uladzislau-zhurauski committed:
Write additional tests of feedback views

as feedback views can now receive images

uladzislau-zhurauski committed:
Write unit tests for managers

uladzislau-zhurauski committed:
Remove inappropriate test db seeding

uladzislau-zhurauski committed:
Move the entire image backend into one view class

uladzislau-zhurauski committed:
Write unit tests for permissions

uladzislau-zhurauski committed:
Fix the link in readme

uladzislau-zhurauski committed:
Merge branch 'OS-10__Write_tests_for_managers' into dev

uladzislau-zhurauski committed:
Merge branch 'OS-11__Write_tests_for_permissions' into dev

uladzislau-zhurauski committed:
Merge branch 'OS-12__Fix_the_link_in_readme' into dev

uladzislau-zhurauski committed:
Add deletion of all feedback images

uladzislau-zhurauski committed:
Write tests for deletion of all feedback images

uladzislau-zhurauski committed:
Remove parser classes from Feedback and Image views

Because these are the default classes.

uladzislau-zhurauski committed:
Add deletion of specific feedback images using PrimaryKeyRelatedField

uladzislau-zhurauski committed:
Add tests for tools

uladzislau-zhurauski committed:
Write tests for deletion of specific feedback images

uladzislau-zhurauski committed:
Modify deletion of specific feedback images using ListField with IntegerField

uladzislau-zhurauski committed:
Rename are_all_elements_in_list and remove prints in feedback views tests

uladzislau-zhurauski committed:
Remove unnecessary argument in FeedbackController.create_feedback

uladzislau-zhurauski committed:
Optimize imports

uladzislau-zhurauski committed:
Remove unnecessary check in validate_images_pk_to_delete

uladzislau-zhurauski committed:
Create IsOwnerOrAdmin factory

uladzislau-zhurauski committed:
Add create_unhandled_value_error to tools

uladzislau-zhurauski committed:
Add address backend

uladzislau-zhurauski committed:
Add tests for address views

uladzislau-zhurauski committed:
Add custom UnhandledValueError exception

uladzislau-zhurauski committed:
Write tests for UnhandledValueError exception

uladzislau-zhurauski committed:
Remove unnecessary import

uladzislau-zhurauski committed:
Add category backend

uladzislau-zhurauski committed:
Use UnhandledValueError in test_image.py

uladzislau-zhurauski committed:
Add tests for category backend

uladzislau-zhurauski committed:
Add category serializer to product serializer, add ProductOutputSerializerForCategory

Merge ProductDetailSerializer and ProductListSerializer into ProductOutputSerializer

uladzislau-zhurauski committed:
Add product material backend

uladzislau-zhurauski committed:
Add nested material serializer to product serializer

uladzislau-zhurauski committed:
Write tests for material backend

uladzislau-zhurauski committed:
Remove unnecessary permission checks in category views

uladzislau-zhurauski committed:
Small refactor in category view tests

uladzislau-zhurauski committed:
Fix method name of ProductMaterialDAL

uladzislau-zhurauski committed:
Add order backend

uladzislau-zhurauski committed:
Write tests for order backend

uladzislau-zhurauski committed:
Add order item backend

uladzislau-zhurauski committed:
Write tests for order item views

uladzislau-zhurauski committed:
Add missing serializer relationships

Create DynamicFieldsModelSerializer to give descendants the ability to remove specific fields from them when they are initialized
Move detail and list serializers into one output serializer for image and feedback serializers
Rename ProductMaterialOutputSerializer to MaterialOutputSerializer for brevity (same for MaterialInputSerializer)
Category serializer now serializes parent and child categories at the same time
Remove RecursiveField as unnecessary

uladzislau-zhurauski committed:
Add content object representation to image output serializer

uladzislau-zhurauski committed:
Add automation to get_content_object

uladzislau-zhurauski committed:
Add user backend

uladzislau-zhurauski committed:
Add tests for user views

uladzislau-zhurauski committed:
Add tests for new IsUserOrAdmin permission

uladzislau-zhurauski committed:
Return dto in compute_superuser_settings()

uladzislau-zhurauski committed:
Rename the product-material-list url to material-list for brevity

uladzislau-zhurauski committed:
Limit allowed methods in all views

uladzislau-zhurauski committed:
Write tests to restrict allowed methods in views

uladzislau-zhurauski committed:
Sort urls alphabetically

uladzislau-zhurauski committed:
Add create, update and delete operations to the product backend

Add new add_and_check_permissions views decorator for convenience

uladzislau-zhurauski committed:
Write tests for create, update and delete operations of the product backend

uladzislau-zhurauski committed:
Rename some functions and arguments

uladzislau-zhurauski committed:
Change the relationship between product and material to many-to-many

uladzislau-zhurauski committed:
Modify product material backend

after changing the relationship between product and material to many-to-many

uladzislau-zhurauski committed:
Modify product material tests

uladzislau-zhurauski committed:
Change removing unneeded input serializer field

uladzislau-zhurauski committed:
Modify product backend

after changing the relationship between product and material to many-to-many

uladzislau-zhurauski committed:
Modify product tests

uladzislau-zhurauski committed:
Improve argument names in ProductMaterialFactory

veranika-zakharchenia commented on shop/controllers/image.py:
```diff
@@ -0,0 +1,41 @@
+            raise Http404
+
+    @classmethod
+    def update_image(cls, image_pk, image, content_type, object_id):
```
why do you get_image, after create_image, after you validate created image and all these actions take place in update method?

veranika-zakharchenia commented on shop/controllers/product.py:
```diff
@@ -0,0 +1,107 @@
+                       materials=None, images=None, images_to_delete=None):
+        product_obj = cls.get_product(product_pk, True)
+        cls.update_product_materials(product_obj, materials)
+        cls.process_images_to_delete(product_obj, images_to_delete)
```
why do you call images_to_delete when it is None?

veranika-zakharchenia commented on shop/controllers/product_material.py:
```diff
@@ -0,0 +1,37 @@
+
+    @classmethod
+    def is_the_same_name(cls, material_pk, new_name):
+        if new_name == cls.get_material(material_pk).name:
```
```suggestion
        return new_name == cls.get_material(material_pk).name:
```

uladzislau-zhurauski replied:
Oops, really

veranika-zakharchenia commented on shop/permissions.py:
```diff
@@ -0,0 +1,45 @@
+    return IsOwnerOrAdmin
+
+
+class IsUserOrAdmin(BasePermission):
```
```suggestion
class PermissionValidator(BasePermission):
```
or checker

uladzislau-zhurauski committed:
Refactor validate_image() method

uladzislau-zhurauski committed:
Refactor update_product() method

uladzislau-zhurauski committed:
Refactor is_the_same_name() method

uladzislau-zhurauski committed:
Rename IsUserOrAdmin to PermissionValidator

veranika-zakharchenia commented on shop/urls.py:
```diff
@@ -0,0 +1,46 @@
+    path('addresses/<int:pk>/', AddressView.as_view(http_method_names=['get', 'put', 'delete']), name='address-detail'),
+    path('auth/', include('rest_framework.urls', namespace='rest_framework')),
+    path('categories/', CategoryView.as_view(http_method_names=['get', 'post']), name='category-list'),
+    path('categories/<int:pk>/', CategoryView.as_view(http_method_names=['get', 'put', 'delete']),
+         name='category-detail'),
+    path('category/<int:category_pk>/', ProductView.as_view(http_method_names=['get']),
+         name='product-list-by-category'),
```
what is the difference between these two endpoints?

uladzislau-zhurauski replied:
`category/<int:category_pk>/` retrieves a list of products filtered by a specific category

veranika-zakharchenia commented on shop/tests/test_views/test_order.py:
```diff
@@ -0,0 +1,107 @@
+    )
+    def test_delete_order(self, client_type, status_code, order_pk, multi_client):
+        url = reverse('order-detail', kwargs={'pk': order_pk})
+        if client_type == ClientType.AUTHOR_CLIENT and order_pk == existent_pk:
```
separate test for each case

veranika-zakharchenia commented on shop/tests/test_views/test_order.py:
```diff
@@ -0,0 +1,107 @@
+        order = Order.objects.first()
+        data = {'address': existent_pk}
+        url = reverse('order-detail', kwargs={'pk': order.pk})
+        user = order.user if client_type == ClientType.AUTHOR_CLIENT else None
```
no conditions in tests

veranika-zakharchenia commented on shop/tests/test_views/test_order.py:
```diff
@@ -0,0 +1,107 @@
+        (ClientType.ADMIN_CLIENT, status.HTTP_200_OK)
+    ])
+    def test_get_order_list(self, client_type, status_code, multi_client):
+        if client_type == ClientType.AUTH_CLIENT:
```
one test must describe one conception

veranika-zakharchenia commented on shop/tests/test_views/test_image.py:
```diff
@@ -0,0 +1,101 @@
+        elif image == Arg.INCORRECT:
+            data['image'] = get_in_memory_file
+
+        if method_type == 'post':
```
the same

veranika-zakharchenia commented on shop/tests/conftest.py:
```diff
@@ -0,0 +1,132 @@
+register(OrderItemFactory)
+register(ProductMaterialFactory)
+
+nonexistent_pk = 0
```
consts should be in upper case

veranika-zakharchenia commented on shop/serializers/image.py:
```diff
@@ -0,0 +1,42 @@
+        for image_model_class in get_image_models():
+            if isinstance(instance.content_object, image_model_class):
+                image_model_module = importlib.import_module(f'{__package__}.{image_model_class.__name__.lower()}')
+                image_model_serializer = getattr(image_model_module, f'{image_model_class.__name__}OutputSerializer')
```
move class name  to variable

uladzislau-zhurauski committed:
Capitalize constants

uladzislau-zhurauski committed:
Refactor get_content_object() of ImageOutputSerializer

uladzislau-zhurauski committed:
Rewrite order view tests

uladzislau-zhurauski committed:
Rewrite order item view tests

uladzislau-zhurauski committed:
Modify multi_client fixture

Raise an error when multi_client fixture is called with author client type and no user object

uladzislau-zhurauski committed:
Rewrite address view tests

uladzislau-zhurauski committed:
Rewrite category view tests

uladzislau-zhurauski committed:
Rewrite feedback view tests

uladzislau-zhurauski committed:
Rewrite image view tests

uladzislau-zhurauski committed:
Rewrite product view tests

uladzislau-zhurauski committed:
Rewrite material view tests

uladzislau-zhurauski committed:
Rewrite user view tests
